[addon] binary add-on library rework for API level system

AlwinEsch opened the pull request:
**This request include a complete rework about the binary add-on library interface from them to kodi.**

It has as mean features and goal:
- Use only C++ on add-on itself to make it more easy and to prevent mem leaks
- Remove of all currently present shared libraries of add-ons
- Handle all calls from add-on to kodi only with "C" handled from independent over function table
- Reduce the dlopen call amount to only two, `Register` and `Unregister`
- Bring in a nearly equal function access to Kodi as on Python
- Add a complete good usable doxygen documentation for C++ and Python
- **And as main feature:** Support about API levels to allow also add-on with older versions

There is still a lot todo and maybe a lot to fix, but by the very big size of this request I want to bring it in and do a improvement with "Fix requests"

---

This request is the first step to change the interface from add-on to Kodi, as next steps in a new request comes rework about the Kodi to Add-on interface to use also the API level design.

---

As last request, if everything is in new level design, comes the add of a executable add-on system to use for everything like on Python.

---

Changes on add-on side are visible here on a request for pvr.vdr.vnsi: https://github.com/kodi-pvr/pvr.vdr.vnsi/pull/74

---

**The current link on http://alwinesch.github.io/ is for testing purpose only!!!**

AlwinEsch commented:
Is the way visible here https://github.com/AlwinEsch/kodi/tree/next-addonlib-rework/xbmc/addons/kodi-addon-dev-kit acceptable?

It use the way with a static library on add-on, for the current add-on's used from **kodi-addon-sharedlibrary**. 

Reason for me to use a static library is to use the same type of headers with exe's and libraries and the actual communication with Kodi is performed by the library.

For this reason, I have also used short code of the static library for exe's.

Since this will be for much more extensive, comes with this request only the support for shared library.

AlwinEsch commented:
Just a node, here is more to see how mutch comes:
https://github.com/AlwinEsch/kodi/tree/thread-crash-protection/xbmc/addons/kodi-addon-dev-kit

FernetMenta commented on xbmc/addons/binary/BinaryAddonManager.cpp:
```diff
@@ -0,0 +1,303 @@
+//  m_Status.Shutdown();
+}
+
+CBinaryAddonManager &CBinaryAddonManager::GetInstance()
```
please stop using this anti pattern. a singleton is at least as evil as a global.

Montellese replied:
TBH I'm not sure if having all kinds of stuff in `CApplication` and then use it through the `g_application` global is any better than a singleton. Yes you have more control over initialization and deinitialization but you can get the same by explicitly initializing and uninitializing a singleton. Actually I prefer to have a direct dependency on the class I'm using instead of everything depending on `CApplication` which in turn depends on all kinds of other stuff. That way the dependencies are less obvious and changing `CApplication` will result in having to re-compile a lot of other theoretically unrelated code.

FernetMenta replied:
The latter can easily solved by adding an interface and implement this by CApplication.

AlwinEsch commented:
@FernetMenta have removed the use of singleton and used a bit the way like on `m_pPlayer(new CApplicationPlayer)` in CApplication. Is this way OK?

AlwinEsch commented:
Want also in the future to move parts from PVRManager and ADSPManager to it e.g. version update checks.

Also for this I have a question about the future, Acceptable to remove singleton on PVR and DSP by use of a way like this?

`g_application.GetBinaryAddonManager()->GetPVRManager()->xxx()`

ksooo commented:
Maybe in the future pvr add-ons can even be implemented as non binary add-ons? Looking at things like this a method GetPVRManager at class BinaryAddonManager makes not too much sense, imo. Same holds for all other add-on types which are today binary only, btw.

AchimTuran commented:
@AlwinEsch 
> `g_application.GetBinaryAddonManager()->GetPVRManager()->xxx()`

I guess this is the global apporach, which is also not very nice. Sit back and try to understand my [CADSPManager](https://github.com/AchimTuran/adsp.template/blob/adsp.template_readd/src/template/AudioDSP/Streams/ADSPStreamManager.hpp#L31-L114) class for my adsp.template, which only uses static methods.

You could also try out the [Monostate NVI Proto-Pattern](http://accu.org/index.php/journals/2085). I have used this e.g. for a logger or an manager object. It's very easy to use if you understand it and work's very similiar to Singleton except you can create more than one objects ;-)

@FernetMenta 
> please stop using this anti pattern. a singleton is at least as evil as a global.

Would my suggestions be a way to go?

AchimTuran commented:
@AlwinEsch 
> It use the way with a static library on add-on, for the current add-on's used from kodi-addon-sharedlibrary.

In my opinion this is the way to go. Using a static library for binary add-ons, which includes the whole APIs. Consequently we compile the code only once, which should speed up our builds. If we use header only libraries we have to compile the included headers every time we include them in a cpp file.

FernetMenta commented:
Keep it simple and don't do too many things at once :)

For the scope of this PR @AlwinEsch 's suggestion is imo good enough.

@ksooo 

> Maybe in the future pvr add-ons can even be implemented as non binary add-ons?

I doubt that this ever happens. Those would be entirely different anyway.

ksooo commented:
>>    Maybe in the future pvr add-ons can even be implemented as non binary add-ons?

> I doubt that this ever happens. Those would be entirely different anyway.

 For me, it feels conceptionally wrong to obtain the PVR Manager from a Binary Addon Mananger. Wrt separation of concerns this is imo not correct. Why tidly couple PVR Mananger to an Addon Manager or even to a Binary Addon Manager? From architectural pov, the fact that PVR clients are realised as addons is imo completely unrelated to PVR manager functionality.

 Well, there are currently some methods with addon related functionalty in CPVRManager but I think those and only those should be ripped off and maybe put to something like PVRAddonManager, derived from GenericAddonManager.

 just my 2 ct.

AlwinEsch commented:
Hi @ksooo, yes the name BinaryAddonManager is a bit wrong, another goal of the libraries on the addon-dev-kit is to make interfaces for languages like java and others and maybe also the python to it.
Was on begin to have a different name about the already present and big file `xbmc/addon/AddonManager.cpp`.

I change `BinaryAddonManager.cpp` to `AddonInterfaceManager.cpp`if OK.

FernetMenta commented:
We won't get this perfect in a single iteration. There is more to rip off PVRManager. I intend to move all methods for stream handling to the inputstream extension point.

akva2 commented:
bit too much code to digest so i'll just ask; does this AddonInterfaceManager manage instances of add-ons as well?

wsnipex commented on Makefile.in:
```diff
@@ -564,6 +572,10 @@ endif
 	  install -m 0644 $$f $(DESTDIR)$(includedir)/@APP_NAME_LC@ ; \
 	done
+	@for d in $(BINDINGS_DIR); do \
+	  install -d $(DESTDIR)$(includedir)/kodi/$$d; \
```
@APP_NAME_LC@, same below

wsnipex commented on Makefile.in:
```diff
@@ -564,6 +572,10 @@ endif
 	done
+	@for d in $(BINDINGS_DIR); do \
+	  install -d $(DESTDIR)$(includedir)/kodi/$$d; \
+	  (find xbmc/addons/kodi-addon-dev-kit/include/kodi/$$d -maxdepth 1 -not -type d -type f -exec install -m 0644 "{}" $(DESTDIR)$(includedir)/kodi/$$d \;) \
```
you shouldn't need a subshell here, so "()" is not needed

AlwinEsch replied:
I bring currently this back. About the "()", it was required about a conflict with the "\;" on end of find call.

Is there a alternative way about it?

AlwinEsch commented:
Can be my last commit reviewed of one who have experience with configure and related Makefile's.

Is about the build of the cmake based dev-kit. It works here but is not perfect :(

wsnipex commented on project/cmake/CMakeLists.txt:
```diff
@@ -350,6 +350,9 @@ if(ENABLE_X11 AND XRANDR_FOUND)
 endif()
 
+# Addon development kit
+add_subdirectory(${CORE_SOURCE_DIR}/xbmc/addons/kodi-addon-dev-kit build/kodi-addon-dev-kit)
```
please add to project/cmake/treedata/common/ instead

AlwinEsch replied:
Thanks, had overlooked that there are also external stuff in it

wsnipex commented on Makefile.in:
```diff
@@ -347,6 +347,7 @@ libaddon: exports
 	$(MAKE) -C lib/addons/library.kodi.peripheral
 	$(MAKE) -C lib/addons/library.xbmc.pvr
 	$(MAKE) -C lib/addons/library.kodi.inputstream
+	PREFIX=$(DESTDIR)$(prefix) $(MAKE) -C xbmc/addons/kodi-addon-dev-kit
```
you should optionally include the top Makefile.include in(-include ...) in the kodi-addon-dev-kit/Makefile
this way you don't need to pass in those variables

wsnipex commented on xbmc/addons/kodi-addon-dev-kit/Makefile:
```diff
@@ -0,0 +1,30 @@
+ifeq ($(CMAKE),)
```
before this line:
-include @abs_top_srcdir@/Makefile.include

wsnipex commented on xbmc/addons/kodi-addon-dev-kit/Makefile:
```diff
@@ -0,0 +1,30 @@
+endif
+
+ifeq ($(PREFIX),)
+  PREFIX = /usr/local
```
shouldn't be needed after adding the include from above.
That is assuming addon-dev-kit is always built with kodi and after configure

wsnipex commented on xbmc/addons/kodi-addon-dev-kit/Makefile:
```diff
@@ -0,0 +1,30 @@
+
+all: .create-kit
+
+.cmake-kit:
```
this target and create-kit should probably be .phony
cmake will take care to not uselessly rebuild stuff.

and deleting the build dir shouldn't be necessary

wsnipex commented on xbmc/addons/kodi-addon-dev-kit/Makefile:
```diff
@@ -0,0 +1,30 @@
+	$(MAKE) -C build clean
+
+distclean:
+	rm -rf build .create-kit .cmake-kit
```
unrelated to this line, in lack of a better place to put it:
This Makefile will be overwritten by cmake if someone ever manually runs cmake in here without creating a build dir

AlwinEsch commented:
Thanks a lot @wsnipex have removed the Makefile complete to prevent problems if cmake is called on the same folder.

And added it direct to Makefile.in of Kodi, then is on dev-kit folder only his related parts. Also the uninstall was on end OK.

wsnipex commented on configure.ac:
```diff
@@ -1767,11 +1767,9 @@ if test "$build_vendor" != "apple" ; then
-  fi
+AC_CHECK_PROG(HAVE_CMAKE,cmake,"yes","no",)
+if test "$HAVE_CMAKE" = "no" ; then
+  AC_MSG_ERROR($missing_program)
```
while you're at it, can you please change it to AC_MSG_ERROR("cmake not found")?

wsnipex commented:
jenkins build this please

AlwinEsch commented:
What is better about this?
```
CMake Error at CMakeLists.txt:3 (cmake_minimum_required):
  CMake 3.0 or higher is required.  You are running version 2.8.7
```

Update CMake to 3.0 for future or reduce the min version on CMakeLists.txt?

akva2 commented:
Update. Other things need 3 already.

-------- Opprinnelig melding --------
Fra: Alwin Esch <notifications@github.com>
Dato: 26.03.2016 16:49 (GMT+01:00)
Til: xbmc/xbmc <xbmc@noreply.github.com>
Ko: Arne Morten Kvarving <Arne.Morten.Kvarving@sintef.no>
Emne: Re: [xbmc] [addon] binary add-on library rework for API level system (#9368)


What is better about this?

CMake Error at CMakeLists.txt:3 (cmake_minimum_required):
  CMake 3.0 or higher is required.  You are running version 2.8.7


Update CMake to 3.0 for future or reduce the min version on CMakeLists.txt?

—
You are receiving this because you commented.
Reply to this email directly or view it on GitHub<https://github.com/xbmc/xbmc/pull/9368#issuecomment-201866004>

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
Have updated the xcode file, but where is the build of the add-on libraries defined on Mac, there is a bit experience missing by me about it there :) Sorry

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please again

AlwinEsch commented:
With a unknown reason for me is always cmake 2.8.7 used on jenkins, on depends becomes 3.3.2 created but seems to be not used, crazy.

@wsnipex @MartijnKaijser have you a idea why jenkins use his own on system?

AlwinEsch commented:
jenkins build this please again

AlwinEsch commented:
Has found the fault. On jenkins configure of Kodi it use the `native` where cmake is present and on the `make` call of Kodi is `native`not present as /bin. I create a fix about it.

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please this please again

wsnipex commented on Makefile.in:
```diff
@@ -349,7 +349,7 @@ libaddon: exports
 addon-dev-kit:
 	mkdir -p xbmc/addons/kodi-addon-dev-kit/build
-	cd xbmc/addons/kodi-addon-dev-kit/build; $(CMAKE) -DCMAKE_INSTALL_PREFIX=$(DESTDIR)$(prefix) ..
+	cd xbmc/addons/kodi-addon-dev-kit/build; cmake -DCMAKE_INSTALL_PREFIX=$(DESTDIR)$(prefix) ..
```
this doesn't look correct. Or must addon-dev-kit be built for native platform instead of target?
When cross compiling, this will build for native/host platform, which seems wrong to me.

AlwinEsch replied:
I am currently trying to find a way to bring up cmake of "native".

Wanted now first be sure that it is found. But that built does not yet what it should be.

wsnipex replied:
this doesn't make any sense. There is no "target cmake", only the one we build in depends, which is "host native", but is configured with the correct toolchain, cflags etc.
see https://github.com/xbmc/xbmc/blob/master/tools/depends/Makefile.include.in#L58
after depends configure, its Makefile.include
You must use this cmake, so $(CMAKE) is correct for cross compiling. For linux standalone(non cross), you just have to check if $(CMAKE) is empty and set it to "cmake" like you had previously in the addon-dev-kit Makefile. Or make the check for cmake in configure do this for you.

AlwinEsch commented on tools/depends/native/config.buildsteps.in:
```diff
@@ -1,5 +1 @@
-export YASM=@prefix@/@tool_dir@/bin/yasm
-export NASM=@prefix@/@tool_dir@/bin/yasm
-export CMAKE=@prefix@/@tool_dir@/bin/cmake
 export PATH="@prefix@/@tool_dir@/bin:$PATH"
```
@wsnipex is this way OK to change the jenkins `PATH` to make sure the from native generated tools are used first.

Thats also the reason why before on jenkins build it has used on Linux the cmake 2.8.7 from his `/usr/bin` where I thought in native is a old version.

wsnipex replied:
you shouldn't need this config.buildsteps.in at all. See comment above.

AlwinEsch commented:
jenkins build this please

AlwinEsch commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "xbmc/addons/kodi-addon-dev-kit"]
+	path = xbmc/addons/kodi-addon-dev-kit
+	url = git://github.com/AlwinEsch/kodi-addon-dev-kit.git
```
OK, here comes a hard part from me. It is to use the `kodi-addon-dev-kit` as part inside kodi itself and as independent repository for add-on developers.

Has searched about ways to let it inside Kodi itself and to perform a git download with a result as indepenendent code like on https://github.com/AlwinEsch/kodi-addon-dev-kit, but nothing found.

Anyone a idea to have it easy alone for developers but still full inside Kodi?

On travis it worked as own repository for add-ons (https://travis-ci.org/xbmc/xbmc/builds/121236644)

akva2 replied:
Uhm cmake is trivial, externalproject drives the entire addon build system. The others are manual work. Just google cmake externalproject.

AlwinEsch replied:
You mean as `externalproject` to build the dev-kit during Jenkins and Travis build? If yes then it is easy with `externalproject` to define to a folder inside Kodi.

But I see the Problem if the add-on is complete created alone without the build on Kodi. There is a way to clone only a part from repository, but the download use the full code inside them and is a bit to long for a add-on build. Also need this way a rework by hand or script. Has found it here http://stackoverflow.com/questions/600079/is-there-any-way-to-clone-a-git-repositorys-sub-directory-only/13738951#13738951 but inacceptable for me.

akva2 replied:
i thought the whole idea of the dev kit was to stick it in its own repository as you have done?

see https://github.com/fetzerch/xbmc/pull/2 

this is how it works in my fork. the kodi build directory supplies an add-on development environment. this environment is currently copied from the main kodi git, but it can as well come from an external git. you then point the add-on build to this build directory where it find the add-on devkit (-DCMAKE_PREFIX_PATH=&lt;kodi_build_dir&gt;/build. the add-on binaries can be mirrored in the kodi build tree, so you can immediately use the produced binary with that kodi (-DKODI_BUILD_DIR=&lt;kodi_build_dir&gt;).

this way, i simply execute 'make' in the add-on and voila.

AlwinEsch replied:
I also not like the use of `.gitmodules`, but want to have on end a complete own development kit, which can used with ways like this.
- As build and system install inside Kodi Itself, to have e.g. on linux a `/usr/lib/kodi-addon-sharedlibrary-api2.a`
- As build with Kodi's add-on steps
- As build if nowhere available on add-on where it becomes handled on his depends

With the last step you can create addon without the use anything else from Kodi.

akva2 replied:
all of this is possible. it's exactly how i always envisioned it working (except my addon-dev-kit was named kodiplatform...)

AlwinEsch replied:
The main problem still is that the code on dev-kit is also needed from build of Kodi itself :(

akva2 replied:
yes, and hence my initial comment. it's trivial for the cmake build system, for the others you have to do it manually supplied through depends on those platforms, through a separate build step as a bat file in visual studio (or through windows depends), manual rule in autotools.

MartijnKaijser replied:
wth? Where not gonna include the awful gitmodule stuff in here again. It's evil!

AlwinEsch replied:
Have yet another idea get, Thank you and Sorry, sometimes I have stupid questions to get new ideas. This external repository way becomes not used. :D

Montellese replied:
And it should be part of the "xbmc" account.

AlwinEsch commented on xbmc/addons/kodi-addon-dev-kit/CMakeLists.txt:
```diff
@@ -2,6 +2,17 @@ project(kodi-addon-dev-kit)
+    COMMENT "Generating API documentation with Doxygen" VERBATIM
+  )
+endif(DOXYGEN_FOUND)
+
```
How is the current documentation of Python on http://mirrors.xbmc.org/docs/python-docs/16.x-jarvis/ generated and uploaded? From hand or somewhere automatic?

Also another question. For what is the on https://github.com/xbmc/xbmc/blob/master/codegenerator.mk#L57-L59 generated doxygen xml needed, is it to use on python itself to have some docs for Kodi? Or is it another way as source for `mirrors.xbmc.org`?

akva2 replied:
It is used by the code generator to supply documentation in the generated python modules.

AlwinEsch replied:
Thanks, you mean to have it here, on e.g. AddonModuleXbmcgui.cpp?

```cpp
  PyDoc_STRVAR(XBMCAddon_xbmcgui_ListItem_setArt__doc__,
               "setArt(values) -- Sets the listitem's art \n"
"\n"
"values : dictionary - pairs of { label: value }.\n"
"\n"
"    - Some default art values (any string possible):\n"
"        - thumb : string - image filename\n"
"        - poster : string - image filename\n"
"        - banner : string - image filename\n"
"        - fanart : string - image filename\n"
"        - clearart : string - image filename\n"
"        - clearlogo : string - image filename\n"
"        - landscape : string - image filename\n"
"        - icon : string - image filename\n"
"\n"
"\n"
"\n"
"example:\n"
"        - self.list.getSelectedItem().setArt({ 'poster': 'poster.png', 'banner' : 'banner.png' }) \n");
```

Good to know then I can check it with the new doxygen documentation style.

akva2 replied:
Yes exactly. It is then exposed as typical python docstrings (of limited value since you cannot run an interactive shell against kodi but still..)

-------- Opprinnelig melding --------
Fra: Alwin Esch <notifications@github.com>
Dato: 07.04.2016 06:08 (GMT+01:00)
Til: xbmc/xbmc <xbmc@noreply.github.com>
Ko: Arne Morten Kvarving <Arne.Morten.Kvarving@sintef.no>
Emne: Re: [xbmc/xbmc] [addon] binary add-on library rework for API level system (#9368)


In xbmc/addons/kodi-addon-dev-kit/CMakeLists.txt<https://github.com/xbmc/xbmc/pull/9368#discussion_r58816987>:

> @@ -2,6 +2,17 @@ project(kodi-addon-dev-kit)
>
>  cmake_minimum_required(VERSION 2.8.7)
>
> +# add a target to generate API documentation with Doxygen
> +find_package(Doxygen)
> +
> +if(DOXYGEN_FOUND)
> +  add_custom_target(doc
> +    ${DOXYGEN_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/doxygen/Doxyfile
> +    WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
> +    COMMENT "Generating API documentation with Doxygen" VERBATIM
> +  )
> +endif(DOXYGEN_FOUND)
> +


Thanks, you mean to have it here, on e.g. AddonModuleXbmcgui.cpp?

  PyDoc_STRVAR(XBMCAddon_xbmcgui_ListItem_setArt__doc__,
               "setArt(values) -- Sets the listitem's art \n"
"\n"
"values : dictionary - pairs of { label: value }.\n"
"\n"
"    - Some default art values (any string possible):\n"
"        - thumb : string - image filename\n"
"        - poster : string - image filename\n"
"        - banner : string - image filename\n"
"        - fanart : string - image filename\n"
"        - clearart : string - image filename\n"
"        - clearlogo : string - image filename\n"
"        - landscape : string - image filename\n"
"        - icon : string - image filename\n"
"\n"
"\n"
"\n"
"example:\n"
"        - self.list.getSelectedItem().setArt({ 'poster': 'poster.png', 'banner' : 'banner.png' }) \n");

Good to know then I can check it with the new doxygen documentation style.

—
You are receiving this because you commented.
Reply to this email directly or view it on GitHub<https://github.com/xbmc/xbmc/pull/9368/files/1bf4fd27e56f068e21971e6b708ac66783bdcaba..cfe6c9ec0bf7660117e01a649fc6f004a430c016#r58816987>

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
Here comes the main feature about this new system.

Have added with last 3 commits a example about a add of a new API level. On them is nothing different as on 2 but good visible what needed.

For changes about new parts is on places of last commit only a new function to add and on headers to add in new level function table. All other functions are used still from older level.

For the add-on side (commit "move API level 2 to 3 for add-on side") I has added some macros to set of Level without text changes. Only whats currently not possible is to set the right include string on code comments, but maybe this can be done in future as creation with something like as example "VFSUtils.hpp.in" an @API_LEVEL@ inside.

The commit "[addons] add old API level 2 for add-on side" is only to have the old. Is separated to holds history available on new level.

AlwinEsch commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/api2/definitions.hpp:
```diff
@@ -0,0 +1,50 @@
+#if defined(BUILD_KODI_ADDON)
+  using namespace V2::KodiAPI;
+  using namespace V2;
+#endif
```
Only for code parts is on this file for new level the number needed to change to new, The rest is done with the Macros.

AlwinEsch commented:
https://github.com/AlwinEsch/kodi/blob/next-addonlib-rework/xbmc/addons/kodi-addon-dev-kit/include/kodi/api2/.internal/AddonLib_internal.hpp#L22-L49

:D

AlwinEsch commented:
To change the text with "api*" for documentation I have added my first python script https://github.com/AlwinEsch/kodi/blob/next-addonlib-rework/xbmc/addons/kodi-addon-dev-kit/tools/rename-apilevel-number.py

With them becomes it easy to change the number after a new level comes.

AlwinEsch committed:
[TEMP] add addon repo to own for test

AlwinEsch commented:
Comes to the end :D

Here on pvr.vdr.vnsi is to see what is changed for Level 2: https://github.com/AlwinEsch/pvr.vdr.vnsi/commit/f26debee13e0d2543274847c18a754e3d394e381

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please again

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
Have updated the description about this request on top. The work is now on end of this and want to bring it in!

**There is still a lot to do and maybe to fix**, but want by ther very big size of this request to bring in like it is now and to do improvement with other pull requests.

Please check and give an OK to merge!

AlwinEsch commented:
@AchimTuran can you also check this way with your add-on sytem rework for DSP and check is acceptable.

AlwinEsch commented:
How a add of new API level becomes done is visible here:
https://github.com/AlwinEsch/kodi/blob/next-addonlib-rework/xbmc/addons/binary/interfaces/api3/Addon/Addon_General.cpp

It still add in his function table the functions of old level, if now one function becomes replaced is this table entry address changed to the new function or added. The table is always defined on the static library of the API level interface for this in https://github.com/AlwinEsch/kodi/blob/next-addonlib-rework/xbmc/addons/kodi-addon-dev-kit/include/kodi/api2/.internal/AddonLib_internal.hpp#L1592-L1620

AlwinEsch commented:
jenkins build this please again

AlwinEsch commented:
jenkins build this please again

tamland commented:
Not sure I like how you have duplicated the entire python documentation. it's more manual copy-paste work for anyone who want to make changes..

tamland commented on xbmc/interfaces/python/Doxyfile:
```diff
@@ -2000,7 +2000,7 @@ INCLUDE_FILE_PATTERNS  =
 # This tag requires that the tag ENABLE_PREPROCESSING is set to YES.
 
-PREDEFINED             =
+PREDEFINED             = SWIG
```
Have you actually tried this? It will generate non-c++ code

AlwinEsch replied:
Yes, this was required to swap to the old documentation of functions.

Here the examples for:
```cpp
#ifndef SWIG /* Part below used for HTML doc and behind #else for python internal docs */
      ///
      /// \ingroup python_InfoTagMusic
      /// @brief Returns the artist from music as string if present.
      ///
      /// @return [string] Music artist
      ///
      ///
      ///------------------------------------------------------------------------
      ///
      ///
      String getArtist();
#else
      /**
       * getArtist() -- returns a string.\n
       */
      String getArtist();
#endif
```

The upper visible here on website and with `#ifndef SWIG` readed there from his doxygen:
http://alwinesch.github.io/a00300.html#ga1b9f565753d008491c59887680bd570e

And for the other selected with this `SWIG` from the other Doxyfile to create this on AddonModuleXbmc.i.cpp:
´´´cpp
  PyDoc_STRVAR(XBMCAddon_xbmc_InfoTagMusic_getArtist__doc__,
               "getArtist() -- returns a string.\n");

  static PyObject* xbmc_XBMCAddon_xbmc_InfoTagMusic_getArtist (PyHolder* self  , PyObject *args, PyObject *kwds  )
  {
    ...
´´´

AlwinEsch commented:
Sorry @tamland I'm not sure to understand right ;)

Do you like or not?

Problem is that the big part for the website is not supported to create the xml for swig's Python documentation.

Maybe with this it can be better optimized for command line helps.

tamland commented:
Yeah, I don't understand what you are trying to do.. You don't mention it in the pr description and the commit message just says 'rework'. There already is doxygen for these files, and a config file to generate it, so why are you duplicating all the documentation in the same files?

AlwinEsch commented:
Is to change a style like this:
- http://mirrors.kodi.tv/docs/python-docs/16.x-jarvis/xbmcgui.html#ListItem-setInfo

to this:
 - http://alwinesch.github.io/a00306.html#gae8ac78b510d6419d3518f1d0c5671550

The new is for me much better human usable as the old and use the same style as on C++. Also becomes on this in future more documentated to help complete for add-on development.

Not sure how much the Python docs from SWIG are used from the developers but for me is a html more usable and for C++ has this already helped me a lot (although most code in c ++ I was).

AlwinEsch commented:
Maybe can be on SWIG part for above changed to
```cpp
      /**
       * setInfo(type, infoLabels) -- Sets the listitem's infoLabels.\n
       * \n
       * For details see documentation on http://alwinesch.github.io/a00306.html#gae8ac78b510d6419d3518f1d0c5671550
       */....
      void setInfo(const char* type, const InfoLabelDict& infoLabels);
```
for all parts.

MartijnKaijser commented:
Absolutely not.

AlwinEsch commented:
No problem :)

My main reason was to have a since years missing usable documentation for C++. But not want to different them on several places, to have something on wiki, something on http://mirrors.kodi.tv/docs/python-docs and then something new on a place somewhere else.

This makes absolutely no sense to me, everything related to create a add-on and for me must be this on only one place!

AlwinEsch commented:
With wiki is the Problem for to hold it then up to date, with also C++ comes very much then there.

MartijnKaijser commented:
It's plain stupid that you duplicate it. Either leave it as is or rip out the pydocs. It's also unacceptable to only add a reference link to some obscure non kodi domain for further reference m

hudokkow commented:
Easy, create a new home for dev docs. I suggest we pick a new subdomain for this.

Suggestions: developers / doxy / doxygen / you-shall-not-read-this / refactoring / oh-crap-what-did-i-got-myself-into . kodi . tv

AlwinEsch commented:
This domain was only a example ;) (and a usable link), And not to see real!!! 

After this come in whe can do something like __https://kodi.tv/addon-development__

AchimTuran commented:
@AlwinEsch
> EDIT: The current link on http://alwinesch.github.io/ is for testing purpose only!!! ;)

For better reviewing please put this in the first message, which you started this PR. For me it wqs clear that this is not the final link, but not for others I guess.

AchimTuran commented:
@MartjinKajiser
> It's also unacceptable to only add a reference link to some obscure non kodi domain for further reference

Doxygen generates HTML, which I guess you know. We could it upload everywhere we want;-)

I have to agree with you. If we duplicate things it's worse. But putting documentation into several parts is also worse.
So far I think it's a good idea to put all documentation into one place.Which I have to agree with @AlwinEsch.

@AlwinEsch sorry but the others are right it's a little bit confusing. The title of this PR says binary add-ons and you try to merge Python specific stuff here.

AchimTuran commented:
I forgot to ask. Is there someone who can review the spellings?

AchimTuran commented on xbmc/interfaces/legacy/InfoTagVideo.h:
```diff
@@ -39,108 +67,495 @@ namespace XBMCAddon
+      String getTagLine();
+#else
       /**
        * getTagLine() - returns a string.
```
Python documentation seems to be very unclair to me.
@AlwinEsch  wrote 
```
/// @brief Get video tag line if available.
///
/// @return [string] Video tag line
```
Current Documentation says:
```
* getTagLine() - returns a string.
```

AlwinEsch commented:
@AchimTuran brought me also a way for the add-on documentation.

To use it like for my test but with a created github account named `https://github.com/kodi-addons` and to generate a html with the link `http://www.kodi-addons.github.io/`

About the current documentation on Python, together with the C++ (which has nearly same functions), want to improve it after it is in Kodi to make it more clear and usable.

tamland commented:
I'm not saying it wouldn't be nice to have all documentation in one place, 

> EDIT: The current link on http://alwinesch.github.io/ is for testing purpose only!!! ;)

but tbh I don't think this makes any sense. This is c++ documentation for the wrapped c++ functions, not python documentation for the generated python modules. For python developers, the signatures in these docs are nonsensical.

AlwinEsch commented:
Have removed much of not needed c++ parts from python parts, it can be more optimized.

Has looked before over the http://mirrors.kodi.tv/docs/python-docs/16.x-jarvis/ where also a lot of not needed parts are in and removed them. Thought it is acceptable.

Only was there on every function the values not present and only set as (...) which is not usable for me.

As example here the
```
bool addDirectoryItem 	( 	int  	handle,
		const String &  	url,
		ListItem *  	listitem,
		bool  	isFolder = false,
		int  	totalItems = 0 
	)
```
file:///home/alwin/Development/11/kodi/xbmc/addons/kodi-addon-dev-kit/docs/html/group__python__xbmcplugin.html#gac48f917e45acddb66f24dd8ebcebf3cc

This const, & and * is not really needed from my side. But can this confuse a script developer?

tamland commented:
>Only was there on every function the values not present and only set as (...) which is not usable for me.

That is because the actual function is `PyObject* addDirectoryItem(PyObject* self, PyObject *args, PyObject *kwds)`. The one you have documented is an internal c++ function called by the generated CPython code.

The full signature is typed out right below in the docstring:
`addDirectoryItem(handle, url, listitem [,isFolder, totalItems])`.

There are many worse than that one. For example: `std::unique_ptr<std::vector<int> > multiselect()`, the return types in python are (as detailed in the docs): a `list` of ints or `None`. & and * is not the main issue, _all_ the type info is nonsensical because it doesn't exist in python.

AlwinEsch commented:
Have added a try about the way of python and visible here:
http://alwinesch.github.io/group__python___dialog.html

What do you think about?

MartijnKaijser commented:
I don't really find this readable
![image](https://cloud.githubusercontent.com/assets/907436/14560874/cd465daa-0312-11e6-9870-fc22c4f8cf06.png)

AlwinEsch commented:
The next try :D

Have added a macro to generate a header

AlwinEsch commented:
Have added as try on two functions here also the python classes before:
- http://alwinesch.github.io/group__python___dialog.html#gad24eabf4c8107d4674c017db09a193ae

What is better, with or without?

garbear commented:
Is this choice between these two?

* current PyDocs (http://mirrors.kodi.tv/docs/python-docs/16.x-jarvis/xbmc.html)
* new dyxogen (http://alwinesch.github.io/group__python___dialog.html#gad24eabf4c8107d4674c017db09a193ae)?

If so, I much prefer the doxygen, it is much neater and conveys more information.

AlwinEsch commented:
Thanks @garbear 

Have changed on last update the webview to use "code text" for function names, looks for me better. Also for the others?

MartijnKaijser commented:
Looks better

AlwinEsch commented:
Function names on begin with or without class names?

AlwinEsch commented:
Another question, better to have a image before or better behind values???

see file:///home/alwin/Development/11/kodi/xbmc/addons/kodi-addon-dev-kit/docs/html/group__python___dialog.html#gaeba4d3278eff78b902a99e63f112a9b9

MartijnKaijser commented:
I don't see the point of adding screenshots

AlwinEsch commented:
Hi @MartijnKaijser (about your edit) thats true :D

Have reworked a bit and the Dialog part now final for me. Also removed this **(Duh!)** ;) and added a bit better text there (from wikipedia).

About the screenshot has moved it as alone to the top to see direct what behind functions. Like on wikipedia, but can also remove it there if not good.

AlwinEsch commented:
Also added now links about older versions:
see http://alwinesch.github.io/usergroup0.html
OK?

AlwinEsch commented on xbmc/addons/kodi-addon-dev-kit/doxygen/DoxygenLayout.xml:
```diff
@@ -19,7 +19,14 @@
+      <tab type="user" url="http://mirrors.kodi.tv/docs/python-docs/13.0-gotham/" title="XBMC 13.x Gotham"/>
+      <tab type="user" url="http://mirrors.kodi.tv/docs/python-docs/12.2-frodo/" title="XBMC 12.x Frodo"/>
+    </tab>
   </navindex>
```
Here is to see how older version links becomes added.

AlwinEsch commented:
Here one of my last questions before I rework everything on Python docs

The use of a function name like this OK?:
- `Function: xbmcaddon.Addon([id]).setSetting(id, value)`
or better to leave it as?:
- `Function: setSetting(id, value)`

Visible here:
- file:///home/alwin/Development/11/kodi/xbmc/addons/kodi-addon-dev-kit/docs/html/group__python__xbmcaddon.html

AlwinEsch commented:
@Montellese haved added reworked documentation to WSGI implementation here on https://github.com/AlwinEsch/kodi/tree/next-addonlib-rework/xbmc/interfaces/legacy/wsgi
but have on expierience with Python to know exact what is right or wrong. Can you review this files?

The current output from them is visible here: http://alwinesch.github.io/group__python__xbmcwsgi.html

Also have reworked all other Python parts to match a better style with right function names on begin.

AlwinEsch committed:
Fix kodi's global doxygen with add of missing macros

AlwinEsch committed:
[addon] add doxygen documentation for skin controls

AlwinEsch committed:
[addon] binary add-on library rework for API level system

AlwinEsch committed:
[addon] add reworked python documentation

AlwinEsch committed:
Add Travis document test

AlwinEsch committed:
Add Travis document test

AlwinEsch commented:
Temporary closed to prevent not needed test build, using currently own travis to test automatic upload of website data
